R/utils.R

```diff
@@ -12,70 +8,112 @@ phantom_run <- function(args, wait = TRUE) {
+#' \file{~/Library/Application Support} is used. On other platforms (such as
+#' Linux), the directory \file{~/bin} is used. If these directories are not
+#' writable, the directory \file{PhantomJS} under the installation directory of
+#' the \pkg{webshot} package will be tried. If this directory still fails, you
+#' will have to install PhantomJS by yourself.
 #' @param version The version number of PhantomJS.
 #' @return \code{NULL} (the executable is written to a system directory).
 #' @export
 install_phantomjs <- function(version = '2.1.1') {
-  if (!is_windows()) {
-    warning('This function is currently for Windows only')
-    return()
-  }
-
-  appdata <- Sys.getenv('APPDATA', NA)
-  if (is.na(appdata)) stop('The environment variable APPDATA is not set')
-  destdir <- file.path(appdata, 'PhantomJS')
-  dir.create(destdir, showWarnings = FALSE)
-
   owd <- setwd(tempdir())
   on.exit(setwd(owd), add = TRUE)
-  zipfile <- sprintf('phantomjs-%s-windows.zip', version)
-  link <- paste0('https://bitbucket.org/ariya/phantomjs/downloads/', zipfile)
-  download.file(link, zipfile, mode = 'wb')
-  utils::unzip(zipfile)
-  zipdir <- sub('.zip$', '', zipfile)
-  file.copy(file.path(zipdir, 'bin', 'phantomjs.exe'), destdir, overwrite = TRUE)
-  message('phantomjs.exe has been installed to ', normalizePath(destdir))
+  base <- 'https://bitbucket.org/ariya/phantomjs/downloads/'
+  if (is_windows()) {
+    zipfile <- sprintf('phantomjs-%s-windows.zip', version)
+    download.file(paste0(base, zipfile), zipfile, mode = 'wb')
+    utils::unzip(zipfile)
+    zipdir <- sub('.zip$', '', zipfile)
+    exec <- file.path(zipdir, 'bin', 'phantomjs.exe')
+  } else if (is_osx()) {
+    zipfile <- sprintf('phantomjs-%s-macosx.zip', version)
+    download.file(paste0(base, zipfile), zipfile, mode = 'wb')
+    utils::unzip(zipfile)
+    zipdir <- sub('.zip$', '', zipfile)
+    exec <- file.path(zipdir, 'bin', 'phantomjs')
+    Sys.chmod(exec, '0755')  # chmod +x
+  } else {
+    zipfile <- sprintf(
+      'phantomjs-%s-linux-%s.tar.bz2', version,
+      if (grepl('64', Sys.info()[['machine']])) 'x86_64' else 'i686'
+    )
+    download.file(paste0(base, zipfile), zipfile, mode = 'wb')
+    utils::untar(zipfile)
+    zipdir <- sub('.tar.bz2$', '', zipfile)
+    exec <- file.path(zipdir, 'bin', 'phantomjs')
+    Sys.chmod(exec, '0755')  # chmod +x
+  }
+  success <- FALSE
+  dirs <- phantom_paths()
+  for (destdir in dirs) {
+    dir.create(destdir, showWarnings = FALSE)
+    success <- file.copy(exec, destdir, overwrite = TRUE)
+    if (success) break
+  }
   unlink(c(zipdir, zipfile), recursive = TRUE)
+  if (!success) stop(
+    'Unable to install PhantomJS to any of these dirs: ',
+    paste(dirs, collapse = ', ')
+  )
+  message('phantomjs.exe has been installed to ', normalizePath(destdir))
   invisible()
 }
 
+# Possible locations of the PhantomJS executable
+phantom_paths <- function() {
```

Comment:
It is because `Sys.getenv('APPDATA', '')` gets `AppData\\Roaming` while the unzip/tempfile is in `AppData\\Local`